keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/ImageService.kt

```diff
@@ -83,110 +83,91 @@ class ImageService(
       }
 
   /**
-   * Returns the latest image that is present in all regions.
-   * Each ami must have tags.
+   * Find the latest properly tagged images in [account] for each of [regions].
+   *
+   * In many cases all the values in the resulting map will be the same [NamedImage] instance, but
+   * this may not be the case if images were baked separately in each region.
+   *
+   * The resulting map will contain no entry for regions where an image is not found. The calling
+   * code must check this if it requires all regions to be present.
    */
-  suspend fun getLatestNamedImageWithAllRegions(packageName: String, account: String, regions: List<String>): NamedImage? {
-    val images = cloudDriverService.namedImages(
-      user = DEFAULT_SERVICE_ACCOUNT,
-      imageName = packageName,
-      account = account
-    )
-
-    val filteredImages = images
-      .filter { it.hasAppVersion }
-      .sortedWith(NamedImageComparator)
-
-    val eliminatedImages = mutableMapOf<String, String>()
-    val image = filteredImages
-      .find {
-        val errors = mutableListOf<String>()
-        // TODO: Frigga and Rocket version parsing are not aligned. We should consolidate.
-        val curAppVersion = try {
-          AppVersion.parseName(it.appVersion)
-        } catch (ex: Exception) {
-          throw SystemException("trying to parse name for version ${it.appVersion} but got an exception", ex)
-        }
-        if (curAppVersion.packageName != packageName) {
-          errors.add("[package name ${curAppVersion.packageName} does not match required package]")
-        }
-        if (!it.accounts.contains(account)) {
-          errors.add("[image is only in accounts ${it.accounts}]")
-        }
-        if (!it.amis.keys.containsAll(regions)) {
-          errors.add("[image is only in regions ${it.amis.keys}]")
-        }
-        if (!tagsExistForAllAmis(it.tagsByImageId)) {
-          errors.add("[image does not have tags for all regions: existing tags ${it.tagsByImageId}]")
-        }
-        if (errors.isEmpty()) {
-          true
-        } else {
-          eliminatedImages[it.imageName] = errors.joinToString(",")
-          false
-        }
+  suspend fun getLatestNamedImageForAppVersionInRegions(
+    appVersion: AppVersion,
+    account: String,
+    regions: Collection<String>
+  ): Map<String, NamedImage> = coroutineScope {
+    regions.associateWith { region ->
+      async {
+        getLatestNamedImageForAppVersionInRegion(
+          appVersion = appVersion,
+          account = account,
+          region = region
+        )
       }
-    log.debug(
-      "Finding latest qualifying named image for $packageName in account $account and regions $regions:\n " +
-        "selected image=${image?.imageName}\n " +
-        "rejected images=${eliminatedImages.map { it.key + ": " + it.value + "\n" }.joinToString("")}"
-    )
-
-    return image
+    }
+      .mapValues { (_, it) -> it.await() }
+      .filterNotNullValues()
   }
```

Comment:
Method to find images in multiple regions. It's just doing a nice little coroutine-ized async call to `getLatestNamedImageForAppVersionInRegion` for each region. Most often the result will be the same for all regions.

The advantage of doing it this way is that the calling code doesn't have to consider what regions are supported by what image, it can just key off the map and not care if they're all the same image, all different ones, or somewhere in between.

I decided to always return whatever regions we _do_ find, rather than returning nothing if we don't find them all as it's less confusing. This just means the calling code needs to check whether all regions are in the result map (if it cares).